src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/service/mapper/PrisonMovementEventsMapper.kt

```diff
@@ -0,0 +1,72 @@
+package uk.gov.justice.digital.hmpps.educationandworkplanapi.app.service.mapper
+
+import org.mapstruct.Mapper
+import org.mapstruct.Mapping
+import org.mapstruct.ValueMapping
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.client.prisonapi.PrisonMovementEvent
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.client.prisonapi.PrisonMovementEvents
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.client.prisonapi.PrisonMovementType
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.resource.mapper.InstantMapper
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.domain.timeline.TimelineEvent
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.domain.timeline.TimelineEventType
+import java.time.Instant
+import java.util.UUID
+
+@Mapper(
+  uses = [
+    InstantMapper::class,
+  ],
+  imports = [
+    Instant::class,
+    UUID::class,
+  ],
+)
+abstract class PrisonMovementEventsMapper {
+  fun toTimelineEvents(prisonMovementEvents: PrisonMovementEvents): List<TimelineEvent> {
+    val timelineEvents = mutableListOf<TimelineEvent>()
+    prisonMovementEvents.prisonBookings.map { booking ->
+      booking.value.forEach { movement ->
```

Comment:
There's some inherent complexity here as we're converting a Map of `<Long, PrisonMovementEvent>` (keyed by `bookingId`) to  a flat List of `TimelineEvents`. Hopefully this is readable, but suggestions welcomed 👍

Reply:
I don't know, and I don't know exactly how to use it (😬 ) , but might `.flatMap` help / be a better choice?
```
prisonMovementEvents.prisonBookings.flatMap { ....
```

Reply:
I've not played with it in the IDE, but it might be something like this:
```
  fun toTimelineEvents(prisonMovementEvents: PrisonMovementEvents): List<TimelineEvent> {
    return prisonMovementEvents.prisonBookings.flatMap { booking ->
      val bookingId = booking.key.toString()
      booking.value.map { movement ->
        toTimelineEvent(
          bookingId = bookingId,
          prisonMovement = movement,
        )
      }
    }

  }
```

Possibly?

As you have it, the `.map` operation on L27 is arguably a misuse of the `.map` function, because it will be returning a list of null which you are doing nothing with. I think you are using `.map` here as a convenient way of iterating through the `prisonBookings` list ?

Reply:
Funnily enough, I used `flatMap` originally, but ended up changing it as it got confusing for reasons I can't recall! (possibly due to the way I was extracting the `bookingId` at the time, but I'm not sure 🤔).

Anyway, I can see that's a better use of the map method and is generally more concise. Nice work 👍